lib/services/shell.dart

```diff
@@ -0,0 +1,234 @@
+    await overlay._controller.requestDismiss(args);
+  }
+
+  @override
+  Future<void> toggleOverlay(
+    String overlayId, {
+    Map<String, dynamic> args = const {},
+  }) async {
+    if (!currentlyShown(overlayId)) {
+      await showOverlay(overlayId, args: args);
+    } else {
+      await dismissOverlay(overlayId, args: args);
+    }
+  }
+
+  @override
+  bool currentlyShown(String overlayId) {
+    final ShellOverlay overlay = overlays.firstWhere((o) => o.id == overlayId);
+    return overlay._controller.showing;
+  }
+
+  @override
+  ValueNotifier<bool> getShowingNotifier(String overlayId) {
+    final ShellOverlay overlay = overlays.firstWhere((o) => o.id == overlayId);
+    return overlay._controller.showingNotifier;
+  }
+
+  @override
+  List<String> get currentlyShownOverlays {
+    final List<String> shownIds = [];
+    for (final ShellOverlay o in overlays) {
+      if (o._controller.showing) shownIds.add(o.id);
+    }
+    return shownIds;
+  }
+
+  @override
+  void dismissEverything() {
+    for (final String id in currentlyShownOverlays) {
+      dismissOverlay(id);
+    }
+    state?.notify();
+    notifyListeners();
+  }
+
+  @override
+  void showInformativeDialog(String title, String message) =>
+      state?.showInformativeDialog(title, message);
+
+  @override
+  FutureOr<void> start() {}
+
+  @override
+  FutureOr<void> stop() {
+    overlays.clear();
+    callbacks.clear();
+    shellStarted = false;
+  }
+}
+
+class ShellOverlayController<T extends ShellOverlayState> {
+  T? _overlay;
+  final ValueNotifier<bool> showingNotifier = ValueNotifier(false);
+
+  bool get showing => showingNotifier.value;
+  set showing(bool value) => showingNotifier.value = value;
+
+  Future<void> requestShow(Map<String, dynamic> args) async {
+    _requireOverlayConnection();
+    await _overlay!.requestShow(args);
+  }
+
+  Future<void> requestDismiss(Map<String, dynamic> args) async {
+    _requireOverlayConnection();
+    await _overlay!.requestDismiss(args);
+  }
+
+  void _requireOverlayConnection() {
+    if (_overlay == null) {
+      throw Exception(
+        "The controller is not connected to any overlay or it had no time to connect yet.",
+      );
+    }
+  }
+}
+
+abstract class ShellOverlay extends StatefulWidget {
+  final String id;
+  final ShellOverlayController _controller = ShellOverlayController();
+
+  ShellOverlay({
+    required this.id,
+    super.key,
+  });
+
+  @override
+  ShellOverlayState createState();
+}
+
+abstract class ShellOverlayState<T extends ShellOverlay> extends State<T>
+    with TickerProviderStateMixin {
+  late final AnimationController animationController = AnimationController(
+    vsync: this,
+    duration: Constants.animationDuration,
+  );
+  ShellOverlayController get controller => widget._controller;
+
+  Animation<double> get animation => CurvedAnimation(
+        parent: animationController,
+        curve: Constants.animationCurve,
```

Comment:
I feel like 150ms is a tad to fast. 200 or 250 feels a lot better. 300 is defo too slow though.

Reply:
we'll need to introduce new animation constants then